src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -3032,43 +3032,92 @@ SynthesizedRecordConstructor addCtor(RecordDeclarationSyntax declWithParameters)
-                if (!memberSignatures.ContainsKey(ctor))
+                members.Add(ctor);
+                return ctor;
+            }
+
+            void addDeconstruct(SynthesizedRecordConstructor ctor, ImmutableArray<PropertySymbol> properties)
+            {
+                var targetMethod = new SignatureOnlyMethodSymbol(
+                    WellKnownMemberNames.DeconstructMethodName,
+                    this,
+                    MethodKind.Ordinary,
+                    Cci.CallingConvention.HasThis,
+                    ImmutableArray<TypeParameterSymbol>.Empty,
+                    ctor.Parameters.SelectAsArray<ParameterSymbol, ParameterSymbol>(param => new SignatureOnlyParameterSymbol(param.TypeWithAnnotations,
+                                                                                                                              ImmutableArray<CustomModifier>.Empty,
+                                                                                                                              isParams: false,
+                                                                                                                              RefKind.Out
+                                                                                                                              )),
+                    RefKind.None,
+                    isInitOnly: false,
+                    TypeWithAnnotations.Create(compilation.GetSpecialType(SpecialType.System_Void)),
+                    ImmutableArray<CustomModifier>.Empty,
+                    ImmutableArray<MethodSymbol>.Empty);
+
+                if (!memberSignatures.TryGetValue(targetMethod, out Symbol? existingDeconstructMethod))
                 {
-                    members.Add(ctor);
+                    members.Add(new SynthesizedRecordDeconstruct(this, ctor, properties, memberOffset: members.Count, diagnostics));
                 }
                 else
                 {
-                    diagnostics.Add(ErrorCode.ERR_DuplicateRecordConstructor, paramList.Location);
-                }
+                    var deconstruct = (MethodSymbol)existingDeconstructMethod;
 
-                return ctor;
-            }
+                    if (deconstruct.DeclaredAccessibility != Accessibility.Public)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_NonPublicAPIInRecord, deconstruct.Locations[0], deconstruct);
+                    }
 
-            void addDeconstruct(ImmutableArray<ParameterSymbol> parameters, ImmutableArray<PropertySymbol> properties)
-            {
-                var ctor = new SynthesizedRecordDeconstruct(this, parameters, properties, memberOffset: members.Count);
-                if (!memberSignatures.ContainsKey(ctor))
-                {
-                    members.Add(ctor);
+                    if (deconstruct.ReturnType.SpecialType != SpecialType.System_Void && !deconstruct.ReturnType.IsErrorType())
```

Comment:
> && !deconstruct.ReturnType.IsErrorType() [](start = 85, length = 41)

nit: if `Deconstruct` returns an error type, then clearly it's not returning `void`. Feels like this check could be removed.

Reply:
> Feels like this check could be removed.

I do not see any harm in keeping it, find it better to follow the pattern and not relying on any special treatment/analysis around special types.

---
In reply to: [456676635](https://github.com/dotnet/roslyn/pull/46053#discussion_r456676635) [](ancestors = 456676635)